llvm/lib/Analysis/ScalarEvolution.cpp

```diff
@@ -3670,9 +3670,12 @@ ScalarEvolution::getAddRecExpr(SmallVectorImpl<const SCEV *> &Operands,
            "SCEVAddRecExpr operand types don't match!");
     assert(!Operands[i]->getType()->isPointerTy() && "Step must be integer");
   }
-  for (unsigned i = 0, e = Operands.size(); i != e; ++i)
+  for (unsigned i = 0, e = Operands.size(); i != e; ++i) {
     assert(isLoopInvariant(Operands[i], L) &&
            "SCEVAddRecExpr operand is not loop-invariant!");
+    assert(properlyDominates(Operands[i], L->getHeader()) &&
```

Comment:
I find it quite surprising that a value defined after L (i.e. whose live range doesn't include L) is consider loop invariant in L.  I wonder if we don't have a whole class of bugs based on that confusion - i.e. beyond this single change.

Do you have any example of a case where we create a SCEV with operands defined after the defining scope of the SCEV?

Reply:
It looks like we have a separate predicate isAvailableAtLoopEntry() which is isLoopInvariant() && properlyDominates(). I would not be surprised if many of our isLoopInvariant() uses should really be isAvailableAtLoopEntry() uses.

I can try changing the isLoopInvariant() definition and see whether anything breaks.

Reply:
Hm, after investigating more, it looks like it might so happen that some loop will get it's Backage taken count computed using cached SCEV value from another loop (which is properly dominating at the time), but then loop unswitch could unswitch this (or some other related) loop making that so that value used in a cached SCEV is no longer dominating the loop which leads to assert. Was unable to reduce the test case, since it proved to be highly dependent on the order BBs are stored inside the Loop structure among other things. Don't think that this PR is buggy, it probably just exposed a bug/potential issue, similar to the one described in https://github.com/llvm/llvm-project/issues/63970 .